services/platform/platform.go

```diff
@@ -111,16 +116,81 @@ func (s *Service) Connect(ctx context.Context, req *platformpb.ConnectRequest) (
 	})
 	if err != nil {
 		s.l.Errorf("Failed to insert SSO details: %s", err)
-		return nil, status.Error(codes.Internal, "Internal server error")
+		return nil, status.Error(codes.Internal, internalServerError)
 	}
 
 	if err := s.UpdateSupervisordConfigurations(ctx); err != nil {
 		s.l.Errorf("Failed to update configuration of grafana after connecting PMM to Portal: %s", err)
-		return nil, status.Error(codes.Internal, "Internal server error")
+		return nil, status.Error(codes.Internal, internalServerError)
 	}
 	return &platformpb.ConnectResponse{}, nil
 }
 
+// Disconnect disconnects a PMM server from the organization created on Percona Portal.
+func (s *Service) Disconnect(ctx context.Context, req *platformpb.DisconnectRequest) (*platformpb.DisconnectResponse, error) {
+	ssoDetails, err := models.GetPerconaSSODetails(ctx, s.db.Querier)
+	if err == nil {
+		return nil, status.Error(codes.Aborted, "PMM server is not connected to Portal")
+	}
+
+	settings, err := models.GetSettings(s.db)
+	if err != nil {
+		s.l.Errorf("Failed to fetch PMM server ID and address: %s", err)
+		return nil, status.Error(codes.Internal, internalServerError)
+	}
+
+	if err := s.CleanSupervisordConfigurations(ctx); err != nil {
+		s.l.Errorf("Failed to clean configuration of grafana during removing PMM from Portal: %s", err)
+		return nil, status.Error(codes.Internal, internalServerError)
+	}
+
+	err = models.DeletePerconaSSODetails(s.db.Querier)
+	if err != nil {
+		s.l.Errorf("Failed to delete SSO details: %s", err)
+		if e := s.UpdateSupervisordConfigurations(ctx); e != nil {
+			s.l.Errorf("%s %s", rollbackFailed, e)
+		}
+		return nil, status.Error(codes.Internal, internalServerError)
+	}
+
+	nCtx, cancel := context.WithTimeout(ctx, platformAPITimeout)
+	defer cancel()
+
+	err = s.disconnect(nCtx, &disconnectPMMParams{
+		PMMServerID: settings.PMMServerID,
+		AccessToken: ssoDetails.AccessToken.AccessToken,
+	})
+	if err != nil {
+		if e := s.UpdateSupervisordConfigurations(ctx); e != nil {
+			s.l.Errorf("%s %s", rollbackFailed, e)
+		}
+		if e := models.InsertPerconaSSODetails(s.db.Querier, &models.PerconaSSODetailsInsert{
+			ClientID:     ssoDetails.ClientID,
+			ClientSecret: ssoDetails.ClientSecret,
+			IssuerURL:    ssoDetails.IssuerURL,
+			Scope:        ssoDetails.Scope,
+		}); e != nil {
+			s.l.Errorf("%s %s", rollbackFailed, e)
+		}
```

Comment:
Just wondering, why inserting again instead of use a transaction and rollback it in case of fail? My concern is that in a general case another insert may lead to the changed internal ids, which means a new record instead of the old one, so it may break some logic that counting on the old ids. 
It may be not the case here, I'm just curious what do you think about it.

Reply:
Another insert into SSO table doesnt break anything since we dont using incremental ID here, content of table will be same. About TX itseft then we need pass DB instead Querier and it require several methods to rewrite. Also not all steps of disconnect are in DB. So it will only affect only one from 3 steps.

Reply:
Should not we first insert and then update configuration?

Reply:
Order changed.